src/back-end/lib/resources/affiliation.ts

```diff
@@ -58,63 +56,58 @@ const resource: Resource = {
-        };
-        // If no existing, active affiliation, create new affiliation with PENDING status
-        if (!existingAffiliation) {
-          if (!permissions.createAffiliation(request.session, user)) {
-            return invalid({
-              permissions: [permissions.ERROR_MESSAGE]
-            });
-          }
-          const validatedMembershipType = validateMembershipType(membershipType);
-          if (validatedMembershipType.tag === 'invalid') {
-            return invalid({
-              membershipType: ['Invalid membership type provided.']
-            });
-          }
-          return valid({ ...validResponseObject, membershipType, membershipStatus: MembershipStatus.Pending });
-        } else {
-          // If existing, active affiliation, create new affiliation with ACTIVE status and updated role
-          if (existingAffiliation.membershipStatus === MembershipStatus.Active) {
-            if (!permissions.updateAffiliation(connection, request.session, organization)) {
+        const validatedUser = await validateUserId(connection, user);
+        const validatedOrganization = await validateOrganizationId(connection, organization);
+        const validatedMembershipType = affiliationValidation.validateMembershipType(membershipType);
+        if (allValid([validatedUser, validatedOrganization, validatedMembershipType])) {
+          const existingAffiliation = await readOneAffiliation(connection, user, organization);
+          if (!existingAffiliation) {
+            if (!permissions.createAffiliation(request.session, user)) {
               return invalid({
                 permissions: [permissions.ERROR_MESSAGE]
               });
             }
-
-            const validatedMembershipType = validateMembershipType(membershipType);
-            if (validatedMembershipType.tag === 'invalid') {
+            // If no existing, active affiliation, create new affiliation with PENDING status
+            return valid({
+              user,
+              organization,
+              membershipType: (validatedMembershipType.value as MembershipType),
+              membershipStatus: MembershipStatus.Pending
+            });
+          } else {
+            if (!permissions.updateAffiliation(connection, request.session, organization)) {
               return invalid({
-                membershipType: ['Invalid membership type provided.']
+                permissions: [permissions.ERROR_MESSAGE]
               });
             }
-            return valid({ ...validResponseObject, membershipType, membershipStatus: MembershipStatus.Active });
+            // If existing, active affiliation, create new affiliation with ACTIVE status and updated role
+            return valid({
+              user,
+              organization,
+              membershipType: (validatedMembershipType.value as MembershipType),
+              membershipStatus: MembershipStatus.Active
+            });
```

Comment:
I don't understand the use-case for this condition? Why would we simply create a second active affiliation for a user? Also, it doesn't look like the existing affiliation is checked to be active per the comment.

Reply:
I may have forgotten in case we discussed this previously :)

Reply:
The idea was to create a new affiliation to replace their active one (if it exists), otherwise to create a new pending one if it doesn't.  I agree something isn't quite right here though, and I'll take another look.

Reply:
Let me know how you go.

Reply:
Simplified it a bit - and added an extra check for pending state.  Still not ideal, I'm wondering if we should use an ADT to separate out the scenarios where want to approve a pending membership, and creating a new membership.  Let me know if I should do this - don't want to create more front-end work unless necessary.